packages/babel-parser/src/parser/expression.js

```diff
@@ -380,6 +380,39 @@ export default class ExpressionParser extends LValParser {
 
         node.right = this.parseExprOpRightExpr(op, prec, noIn);
 
+        /* this check is for all ?? operators
+         * a ?? b && c for this example
+         * b && c => This is considered as a logical expression in the ast tree
+         * a => Identifier
+         * so for ?? operator we need to check in this case the right expression to have parenthesis
+         * second case a && b ?? c
+         * here a && b => This is considered as a logical expression in the ast tree
+         * c => identifer
+         * so now here for ?? operator we need to check the left expression to have parenthesis
+         * if the parenthesis is missing we raise an error and throw it
+         */
+        if (op === tt.nullishCoalescing) {
+          if (
+            left.type === "LogicalExpression" &&
+            left.operator !== "??" &&
+            !(left.extra && left.extra.parenthesized)
+          ) {
+            throw this.raise(
+              left.start,
+              `Nullish coalescing operator(??) requires parens when mixing with logical operators`,
+            );
+          } else if (
+            node.right.type === "LogicalExpression" &&
+            node.right.operator !== "??" &&
+            !(node.right.extra && node.right.extra.parenthesized)
+          ) {
+            throw this.raise(
+              node.right.start,
+              `Nullish coalescing operator(??) requires parens when mixing with logical operators`,
+            );
+          }
```

Comment:
@jridgewell @vivek12345 Can I ask why you are checking the AST node property? It should be enough to use current token as a func arg and validate against it. That way you 1) improve performance 2) skip duplicate code 3) simplify

This will also improve the performance.

I did it that way when I implemented similar in [my own parser Meriyah](https://github.com/meriyah/meriyah/blob/master/src/parser.ts#L3268-L3270) a

Reply:
It would only work for `a && b ?? c`, but not for `a ?? b && c` since in the second case the precedence is `a ?? (b && c)`

Reply:
It would only work for `a && b ?? c`, but not for `a ?? b && c` since in the second case the precedence is `a ?? (b && c)`

Reply:
@nicolo-ribaudo I'm confused because it will work both lhs and rhs.  I implemented it that way I described. See [here](https://github.com/meriyah/meriyah/blob/master/src/parser.ts#L3406) . And it works very well to if you try it in the [REPL](https://meriyah.github.io/meriyah/)